File: backend/seed.js
require('dotenv').config();
const mongoose = require('mongoose');
const bcrypt = require('bcryptjs');
const User = require('./models/User');
const RFQ = require('./models/RFQ');
const Bid = require('./models/Bid');

const MONGO_URI = process.env.MONGO_URI || 'mongodb://127.0.0.1:27017/british-auction';

const seedDatabase = async () => {
  await mongoose.connect(MONGO_URI);
  console.log('[SEED] Connected to MongoDB');

  // Clear existing data
  await Bid.deleteMany({});
  await RFQ.deleteMany({});
  await User.deleteMany({});
  console.log('[SEED] Cleared existing data');

  // ─── Users ─────────────────────────────────────────────────────────────────
  const password = 'Test@1234';
  const salt = await bcrypt.genSalt(12);
  const hashedPassword = await bcrypt.hash(password, salt);

  const usersData = [
    { name: 'Alice Buyer', email: 'buyer1@test.com', password: hashedPassword, role: 'buyer' },
    { name: 'Bob Buyer', email: 'buyer2@test.com', password: hashedPassword, role: 'buyer' },
    { name: 'Supply Co Ltd', email: 'supplier1@test.com', password: hashedPassword, role: 'supplier' },
    { name: 'Global Freight Inc', email: 'supplier2@test.com', password: hashedPassword, role: 'supplier' },
    { name: 'FastMove Logistics', email: 'supplier3@test.com', password: hashedPassword, role: 'supplier' },
    { name: 'BlueLine Carriers', email: 'supplier4@test.com', password: hashedPassword, role: 'supplier' },
  ];

  // Bypass pre-save hook for seeding (password is already hashed)
  const users = await User.insertMany(usersData);
  console.log(`[SEED] Created ${users.length} users`);

  const buyer1 = users.find((u) => u.email === 'buyer1@test.com');
  const buyer2 = users.find((u) => u.email === 'buyer2@test.com');
  const supplier1 = users.find((u) => u.email === 'supplier1@test.com');
  const supplier2 = users.find((u) => u.email === 'supplier2@test.com');
  const supplier3 = users.find((u) => u.email === 'supplier3@test.com');
  const supplier4 = users.find((u) => u.email === 'supplier4@test.com');

  // ─── RFQs ──────────────────────────────────────────────────────────────────
  const now = new Date();
  const twoHoursFromNow = new Date(now.getTime() + 2 * 60 * 60 * 1000);
  const threeHoursFromNow = new Date(now.getTime() + 3 * 60 * 60 * 1000);
  const fiveHoursFromNow = new Date(now.getTime() + 5 * 60 * 60 * 1000);
  const yesterday = new Date(now.getTime() - 24 * 60 * 60 * 1000);
  const twoDaysAgo = new Date(now.getTime() - 2 * 24 * 60 * 60 * 1000);
  const threeDaysAgo = new Date(now.getTime() - 3 * 24 * 60 * 60 * 1000);
  const pickupDate1 = new Date(now.getTime() + 7 * 24 * 60 * 60 * 1000);
  const pickupDate2 = new Date(now.getTime() + 14 * 24 * 60 * 60 * 1000);

  const rfqsData = [
    {
      referenceId: `RFQ-${now.getFullYear()}-001`,
      name: 'Mumbai to Delhi Freight',
      buyer: buyer1._id,
      bidStartTime: new Date(now.getTime() - 30 * 60 * 1000), // 30 min ago
      bidCloseTime: twoHoursFromNow,
      forcedCloseTime: threeHoursFromNow,
      pickupDate: pickupDate1,
      status: 'active',
      auctionConfig: {
        triggerWindowMinutes: 10,
        extensionDurationMinutes: 5,
        extensionTrigger: 'l1_rank_change',
      },
      extensionLogs: [
        {
          previousCloseTime: new Date(now.getTime() + 1.5 * 60 * 60 * 1000),
          newCloseTime: twoHoursFromNow,
          triggerReason: 'L1 (lowest bidder) changed within trigger window',
          extendedAt: new Date(now.getTime() - 10 * 60 * 1000),
        },
      ],
    },
    {
      referenceId: `RFQ-${now.getFullYear()}-002`,
      name: 'Pune Export Shipment',
      buyer: buyer2._id,
      bidStartTime: threeDaysAgo,
      bidCloseTime: twoDaysAgo,
      forcedCloseTime: yesterday,
      pickupDate: pickupDate2,
      status: 'closed',
      auctionConfig: {
        triggerWindowMinutes: 15,
        extensionDurationMinutes: 10,
        extensionTrigger: 'bid_received',
      },
      extensionLogs: [],
    },
  ];

  const rfqs = await RFQ.insertMany(rfqsData);
  console.log(`[SEED] Created ${rfqs.length} RFQs`);

  const activeRFQ = rfqs.find((r) => r.referenceId === `RFQ-${now.getFullYear()}-001`);

  // ─── Bids (for active RFQ) ─────────────────────────────────────────────────
  // Six bids from four suppliers with varying amounts to show ranking
  const bidsData = [
    {
      rfq: activeRFQ._id,
      supplier: supplier1._id,
      carrierName: 'DTDC Express',
      freightCharges: 12000,
      originCharges: 1500,
      destinationCharges: 1000,
      totalAmount: 14500,
      transitTime: 3,
      quoteValidityDate: fiveHoursFromNow,
      rank: 1,
      isLatest: true,
    },
    {
      rfq: activeRFQ._id,
      supplier: supplier2._id,
      carrierName: 'Blue Dart',
      freightCharges: 13500,
      originCharges: 1200,
      destinationCharges: 800,
      totalAmount: 15500,
      transitTime: 2,
      quoteValidityDate: fiveHoursFromNow,
      rank: 2,
      isLatest: true,
    },
    {
      rfq: activeRFQ._id,
      supplier: supplier3._id,
      carrierName: 'FedEx India',
      freightCharges: 14000,
      originCharges: 2000,
      destinationCharges: 1200,
      totalAmount: 17200,
      transitTime: 4,
      quoteValidityDate: fiveHoursFromNow,
      rank: 3,
      isLatest: true,
    },
    {
      rfq: activeRFQ._id,
      supplier: supplier4._id,
      carrierName: 'Gati Express',
      freightCharges: 15000,
      originCharges: 1800,
      destinationCharges: 1500,
      totalAmount: 18300,
      transitTime: 5,
      quoteValidityDate: fiveHoursFromNow,
      rank: 4,
      isLatest: true,
    },
    // Stale bids (isLatest=false) showing bid history for suppliers 1 & 2
    {
      rfq: activeRFQ._id,
      supplier: supplier1._id,
      carrierName: 'DTDC Express',
      freightCharges: 16000,
      originCharges: 1600,
      destinationCharges: 1100,
      totalAmount: 18700,
      transitTime: 3,
      quoteValidityDate: fiveHoursFromNow,
      rank: null,
      isLatest: false,
    },
    {
      rfq: activeRFQ._id,
      supplier: supplier2._id,
      carrierName: 'Blue Dart',
      freightCharges: 17000,
      originCharges: 1400,
      destinationCharges: 900,
      totalAmount: 19300,
      transitTime: 2,
      quoteValidityDate: fiveHoursFromNow,
      rank: null,
      isLatest: false,
    },
  ];

  await Bid.insertMany(bidsData);
  console.log(`[SEED] Created ${bidsData.length} bids`);

  console.log('\n[SEED] ✅ Database seeded successfully!\n');
  console.log('─────────────────────────────────────');
  console.log('Test Credentials (password: Test@1234)');
  console.log('─────────────────────────────────────');
  console.log('Buyers:    buyer1@test.com | buyer2@test.com');
  console.log('Suppliers: supplier1@test.com | supplier2@test.com');
  console.log('           supplier3@test.com | supplier4@test.com');
  console.log('─────────────────────────────────────\n');

  await mongoose.disconnect();
  process.exit(0);
};

seedDatabase().catch((error) => {
  console.error('[SEED] Error:', error.message);
  process.exit(1);
});

File: backend/app/seed.py
"""
Database seed script — populates test data for local development.

Usage:
    cd backend
    python -m app.seed

Creates:
  - 2 buyer accounts
  - 4 supplier accounts
  - 2 RFQs (one active, one closed)
  - Sample bids on the active RFQ

All test accounts use the password: Test@1234
"""

import asyncio
from datetime import datetime, timedelta, timezone

from passlib.context import CryptContext
from bson import ObjectId

from app.db import users_collection, rfqs_collection, bids_collection

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto", bcrypt__rounds=12)
PASSWORD = pwd_context.hash("Test@1234")


async def seed():
    print("Dropping existing collections...")
    await users_collection.drop()
    await rfqs_collection.drop()
    await bids_collection.drop()

    now = datetime.now(timezone.utc)

    # ----- Users -----
    buyers = [
        {"_id": ObjectId(), "name": "Alice Buyer", "email": "buyer1@test.com", "password": PASSWORD, "role": "buyer", "createdAt": now, "updatedAt": now},
        {"_id": ObjectId(), "name": "Bob Buyer", "email": "buyer2@test.com", "password": PASSWORD, "role": "buyer", "createdAt": now, "updatedAt": now},
    ]
    suppliers = [
        {"_id": ObjectId(), "name": "Supplier One", "email": "supplier1@test.com", "password": PASSWORD, "role": "supplier", "createdAt": now, "updatedAt": now},
        {"_id": ObjectId(), "name": "Supplier Two", "email": "supplier2@test.com", "password": PASSWORD, "role": "supplier", "createdAt": now, "updatedAt": now},
        {"_id": ObjectId(), "name": "Supplier Three", "email": "supplier3@test.com", "password": PASSWORD, "role": "supplier", "createdAt": now, "updatedAt": now},
        {"_id": ObjectId(), "name": "Supplier Four", "email": "supplier4@test.com", "password": PASSWORD, "role": "supplier", "createdAt": now, "updatedAt": now},
    ]

    all_users = buyers + suppliers
    await users_collection.insert_many(all_users)
    print(f"  Inserted {len(all_users)} users")

    # ----- RFQs -----
    active_rfq = {
        "_id": ObjectId(),
        "referenceId": f"RFQ-{now.year}-001",
        "name": "Mumbai → Delhi Freight",
        "buyer": buyers[0]["_id"],
        "bidStartTime": now - timedelta(hours=1),
        "bidCloseTime": now + timedelta(hours=2),
        "forcedCloseTime": now + timedelta(hours=4),
        "pickupDate": now + timedelta(days=3),
        "status": "active",
        "auctionConfig": {
            "triggerWindowMinutes": 15,
            "extensionDurationMinutes": 10,
            "extensionTrigger": "l1_rank_change",
        },
        "extensionLogs": [],
        "createdAt": now - timedelta(hours=2),
        "updatedAt": now,
    }

    closed_rfq = {
        "_id": ObjectId(),
        "referenceId": f"RFQ-{now.year}-002",
        "name": "Chennai → Bangalore Freight",
        "buyer": buyers[1]["_id"],
        "bidStartTime": now - timedelta(days=2),
        "bidCloseTime": now - timedelta(days=1),
        "forcedCloseTime": now - timedelta(hours=12),
        "pickupDate": now + timedelta(days=1),
        "status": "closed",
        "auctionConfig": {
            "triggerWindowMinutes": 10,
            "extensionDurationMinutes": 5,
            "extensionTrigger": "bid_received",
        },
        "extensionLogs": [],
        "createdAt": now - timedelta(days=3),
        "updatedAt": now - timedelta(days=1),
    }

    await rfqs_collection.insert_many([active_rfq, closed_rfq])
    print("  Inserted 2 RFQs")

    # ----- Bids on the active RFQ -----
    bids = [
        {
            "rfq": active_rfq["_id"],
            "supplier": suppliers[0]["_id"],
            "carrierName": "FastShip Logistics",
            "freightCharges": 15000.0,
            "originCharges": 2000.0,
            "destinationCharges": 1500.0,
            "totalAmount": 18500.0,
            "transitTime": 3,
            "quoteValidityDate": now + timedelta(days=7),
            "isLatest": True,
            "rank": 1,
            "submittedAt": now - timedelta(minutes=30),
            "createdAt": now - timedelta(minutes=30),
            "updatedAt": now - timedelta(minutes=30),
        },
        {
            "rfq": active_rfq["_id"],
            "supplier": suppliers[1]["_id"],
            "carrierName": "QuickHaul Transport",
            "freightCharges": 16000.0,
            "originCharges": 2500.0,
            "destinationCharges": 1800.0,
            "totalAmount": 20300.0,
            "transitTime": 2,
            "quoteValidityDate": now + timedelta(days=7),
            "isLatest": True,
            "rank": 2,
            "submittedAt": now - timedelta(minutes=20),
            "createdAt": now - timedelta(minutes=20),
            "updatedAt": now - timedelta(minutes=20),
        },
        {
            "rfq": active_rfq["_id"],
            "supplier": suppliers[2]["_id"],
            "carrierName": "EcoMove Carriers",
            "freightCharges": 17500.0,
            "originCharges": 2200.0,
            "destinationCharges": 2000.0,
            "totalAmount": 21700.0,
            "transitTime": 4,
            "quoteValidityDate": now + timedelta(days=5),
            "isLatest": True,
            "rank": 3,
            "submittedAt": now - timedelta(minutes=10),
            "createdAt": now - timedelta(minutes=10),
            "updatedAt": now - timedelta(minutes=10),
        },
    ]

    await bids_collection.insert_many(bids)
    print(f"  Inserted {len(bids)} bids")

    print("\nSeed complete!")
    print("  Buyers:    buyer1@test.com, buyer2@test.com")
    print("  Suppliers: supplier1@test.com .. supplier4@test.com")
    print("  Password:  Test@1234")


if __name__ == "__main__":
    asyncio.run(seed())
